node/src/components/consensus/protocols/highway/tests/consensus_environment.rs

```diff
@@ -0,0 +1,434 @@
+                        None,
+                    )
+                })
+                .collect();
+            for unit in slow_witness_units {
+                let highway_msg = HighwayMessage::NewVertex(Vertex::Unit(unit));
+                let msg = SerializedMessage::from_message(&highway_msg);
+                self.handle_message(msg, min_round_len * 3 / 4);
+            }
+            self.add_vertices((self.current_round_start + min_round_len * 3 / 4 + 1).into());
+        };
+
+        self.current_round_start = self.current_round_start.saturating_add(min_round_len);
+    }
+
+    fn this_node_propose(&mut self) {
+        if (self.current_round_start / self.min_round_len()).trailing_zeros()
+            < self.our_round_exp() as u32
+        {
+            // we're not proposing if we shouldn't even be participating in this round
+            return;
+        }
+
+        let now: Timestamp = self.current_round_start.into();
+        // the timer triggers a request for block content
+        let outcomes =
+            self.highway
+                .handle_timer(now, now, TIMER_ID_ACTIVE_VALIDATOR, &mut self.rng);
+        self.finalize_blocks(&outcomes);
+        // the request contains necessary block context - extract it
+        let block_context = outcomes
+            .iter()
+            .find_map(|outcome| match outcome {
+                ProtocolOutcome::CreateNewBlock(context) => Some(context.clone()),
+                _ => None,
+            })
+            .unwrap_or_else(|| {
+                self.dump();
+                panic!("outcomes didn't contain CreateNewBlock: {:?}", outcomes)
+            });
+        // this should create the proposal unit, add it to the state and create a message to be
+        // broadcast - we can ignore the message, because we don't keep other consensus
+        // instances
+        let outcomes = self
+            .highway
+            .propose(ProposedBlock::new(Default::default(), block_context), now);
+        self.finalize_blocks(&outcomes);
+    }
+
+    fn create_swunit_from(
+        &self,
+        creator: ValidatorIndex,
+        keypair: &Keypair,
+        state: &State<ClContext>,
+        delay: u64,
+        value: Option<<ClContext as Context>::ConsensusValue>,
+    ) -> SignedWireUnit<ClContext> {
+        let seq_number = {
+            let prev_unit_hash = state.panorama().get(creator).unwrap().correct();
+            prev_unit_hash.map_or(0, |hash| state.unit(hash).seq_number.saturating_add(1))
+        };
+        let wunit: WireUnit<ClContext> = WireUnit {
+            panorama: state.panorama().clone(),
+            creator,
+            instance_id: *self.highway().instance_id(),
+            value,
+            seq_number,
+            timestamp: (self.current_round_start + delay).into(),
+            round_exp: 0,
+            endorsed: BTreeSet::new(),
+        };
+        SignedWireUnit::new(wunit.into_hashed(), keypair)
+    }
+
+    fn other_node_propose(
```

Comment:
This is where my conceptual model of this starts to unravel. The conceit here seems to be that we track a limited sort of Highway state that's limited to the 0th validator. So, in this function, we can only get the state to base _other_ nodes' proposals on from the 0th validator. However, some of these nodes are slow and don't see the latest (among states possible across all nodes, if this were full distributed Highway) state. This seems weird, especially since, above, they do create witnesses based on what amounts to their subjective view of the state. Concisely, my concern here is that we're no longer modeling Highway, because this is Highway with a special omniscience that _only_ kicks in when a node is proposing (alternatively, it restricts us to modeling situations where nodes always "catch up" after performing poorly in a particular round, before anything else is proposed).

Reply:
The idea here is that we only model the state held by 0th node, and the 0th node processing messages received from other validators. The state held by the other validators is assumed to be the same as the 0th node's state at some point in time - and which point it is, depends on whether the other validator is slow or not.

Maybe it would be simpler to just simulate multiple Highway instances and messages between them, and then add delays to messages sent from and received by particular validators... I wanted to keep this minimal, but it's possible that it complicates the logic unnecessarily.